text/0000-nonparametric-dropck.md

```diff
@@ -0,0 +1,541 @@
+broad as it needs to be (and no broader).
+
+## "Sort-of Guarded" Escape Hatch
+
+We could add the escape hatch but continue employing the current
+dropck analysis to it. This would essentially mean that code would have
+to apply the unsafe attribute to be considered for parametricity, but
+if there were obvious problems (namely, if the type parameter had a trait bound)
+then the attempt to opt into parametricity would be ignored and the
+strict ordering restrictions on the lifetimes would be imposed.
+
+I only mention this because it occurred to me in passing; I do not
+really think it has much of a benefit. It would potentially lead
+someone to think that their code has been proven sound (since the
+`dropck` would catch some mistakes in programmer reasoning) but the
+pitfalls with respect to specialization would remain.
+
+## Continue Supporting Parametricity
+[continue supporting parametricity]: #continue-supporting-parametricity
+There may be ways to revise the language so that functions can declare
+that they must be parametric with respect to their type parameters.
+Here we sketch two potential ideas for how one might do this, mostly to
+give a hint of why this is not a trivial change to the language.
+
+Neither design is likely to be adopted, at least as described here,
+because both of them impose significant burdens on implementors of
+parametric destructors, as we will see.
+
+(Also, if we go down this path, we will need to fix other bugs in the
+Drop Check rule, where, as previously noted, parametricity is a
+[necessary but *insufficient* condition][parametricity-insufficient] for soundness.)
+
+### Parametricity via effect-system attributes
+
+One feature of the [impl specialization] RFC is that all functions that
+can be specialized must be declared as such, via the `default` keyword.
+
+This leads us to one way that a function could declare that its body
+must not be allows to call into specialized methods: an attribute like
+`#[unspecialized]`. The `#[unspecialized]` attribute, when applied to
+a function `fn foo()`, would mean two things:
+
+ * `foo` is not allowed to call any functions that have the `default` keyword.
+
+ * `foo` is only allowed to call functions that are also marked `#[unspecialized]`
+
+All `fn drop` methods would be required to be `#[unspecialized]`.
+
+It is the second bullet that makes this an ad-hoc effect system: it provides
+a recursive property ensuring that during the extent of the call to `foo`,
+we will never invoke a function marked as `default` (and therefore, I *think*,
+will never even potentially invoke a method that has been specialized).
+
+It is also this second bullet that represents a signficant burden on
+the destructor implementor. In particular, it immediately rules out
+using any library routine unless that routine has been marked as
+`#[unspecialized]`. The attribute is unlikely to be included on any
+function unless the its developer is making a destructor that calls it
+in tandem.
+
```

Comment:
I feel like the effect checking is simple enough that the compiler could do it without an annotation - something equivalent to `unspecialised` would need to be saved into the metadata, otherwise it seems like a fairly simple (though global) analysis.